ext/storage_sources/gcp_store/gcp_storage_source.cpp

```diff
@@ -294,60 +313,103 @@ gcp_file_open(WT_FILE_SYSTEM *file_system, WT_SESSION *session, const char *name
+    if (ret != 0) {
+        std::cerr << "gcp_object_size: GetObjectMetadata request to google cloud failed."
+                  << std::endl;
+        return (ret);
+    }
+    *sizep = size;
 
     return 0;
 }
 
 static int
-gcp_file_size(WT_FILE_SYSTEM *file_system, WT_SESSION *session, const char *name, wt_off_t *sizep)
+gcp_object_list_helper(WT_FILE_SYSTEM *file_system, WT_SESSION *session, const char *directory,
+  const char *prefix, char ***object_list, uint32_t *count, bool list_single)
 {
-    WT_UNUSED(file_system);
-    WT_UNUSED(session);
-    WT_UNUSED(name);
-    WT_UNUSED(sizep);
+    gcp_file_system *fs = reinterpret_cast<gcp_file_system *>(file_system);
+    gcp_store *gcp = reinterpret_cast<gcp_file_system *>(fs)->storage_source;
+    std::vector<std::string> objects;
+    std::string complete_prefix;
 
-    return 0;
+    *count = 0;
+
+    if (directory != nullptr) {
+        complete_prefix += directory;
+        if (complete_prefix.length() > 1 && complete_prefix.back() != '/')
+            complete_prefix += '/';
+    }
+
+    if (prefix != nullptr)
+        complete_prefix += prefix;
+
+    int ret;
+
+    ret = list_single ? fs->gcp_conn->list_objects(complete_prefix, objects, true) :
+                        fs->gcp_conn->list_objects(complete_prefix, objects, false);
+
+    if (ret != 0) {
+        std::cerr << "gcp_object_list: ListObjects request to google cloud failed." << std::endl;
+        return (ret);
+    }
+    *count = objects.size();
+
+    std::cerr << "gcp_object_list: ListObjects request to google cloud succeeded. Received " +
+        std::to_string(*count) + " objects."
+              << std::endl;
+    gcp_object_list_add(*gcp, object_list, objects, *count);
+
+    return ret;
 }
 
 static int
 gcp_object_list(WT_FILE_SYSTEM *file_system, WT_SESSION *session, const char *directory,
   const char *prefix, char ***object_list, uint32_t *count)
 {
-    WT_UNUSED(file_system);
-    WT_UNUSED(session);
-    WT_UNUSED(directory);
-    WT_UNUSED(prefix);
-    WT_UNUSED(object_list);
-    WT_UNUSED(count);
-
-    return 0;
+    return gcp_object_list_helper(
+      file_system, session, directory, prefix, object_list, count, false);
 }
 
 static int
 gcp_object_list_add(const gcp_store &gcp_, char ***object_list,
   const std::vector<std::string> &objects, const uint32_t count)
 {
-    WT_UNUSED(gcp_);
-    WT_UNUSED(object_list);
-    WT_UNUSED(objects);
-    WT_UNUSED(count);
+    char **entries;
+    if ((entries = reinterpret_cast<char **>(malloc(sizeof(char *) * count))) == nullptr) {
+        std::cerr << "gcp_object_list_add: unable to allocate memory for object list." << std::endl;
+        return ENOMEM;
+    }
+
+    for (int i = 0; i < count; i++) {
+        if ((entries[i] = strdup(objects[i].c_str())) == nullptr) {
+            std::cerr << "gcp_object_list_add: unable to allocate memory for object string."
+                      << std::endl;
+            return ENOMEM;
```

Comment:
Memory leak: `entries` + (i * strdup).

Standard defines `malloc(0)` as undefined https://en.cppreference.com/w/c/memory/malloc . I think most sane implementations return `nullptr` but best to avoid. 

To ensure cleanup it might be easier to build the array of `char*` using a temporary `std::vector`.

Possibly something like:
```
// If count of zero is invalid argument can replace with an assertion.
if (count < 1) {
  *object_list = nullptr;
  return 0;
}
size_t entries_size = sizeof(char*) * count;
char **entries = malloc(entries_size);
std::vector<char *> tmp(count);
uint32_t copied = 0;
// TODO strdup() copying loop copied < count
if (copied < count) {
  // It is possible to use std::for_each() here as free(nullptr) does nothing but if count is large...
  std::for_each_n(tmp.begin(), copied, [](char *p){ free(p); }); // there may be a simpler way to invoke free
  return ENOMEM;
}
std::copy(tmp.begin(), tmp.end(), std::begin(entries));
return 0;
```

Reply:
In retrospect the temporary vector doesn't really add much except 2 * N the memory used, using the allocated `entries` array is better. `std::for_each_n` can still be used, ie `std::begin(entries)`.

Reply:
Hi Marc, can you explain how this would be done without the tmp vector?

Reply:
Neither compiled nor tested:
```suggest
if (count < 1) {
    *object_list = nullptr;
     return 0;
}

char **entries;
if ((entries = reinterpret_cast<char**>(malloc(sizeof(char *) * count))) == nullptr)
     return ENOMEM;
uint32_t copied;
for (copied = 0; copied < count; copied++) {
    if ((entries[i] = strdup(objects[i].c_str())) == nullptr)
        break;
}

if (copied < count) {
    uint32_t i;
    for (i = 0; i < count; i++)
        free(entries[i]);
    free(entries);
    return ENOMEM;
}

*object_list = entries;
return 0;
```

Reply:
updated d11e042e49654d91f7ee8bd8450515ee77e1766e